packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,203 @@
+        public
+        payable
+        nonReentrant
+    {
+        // Do all fast withdrawals
+        for (uint i = 0; i < fastWithdrawals.length; i++) {
+            executeFastWithdrawal(fastWithdrawals[i]);
+        }
+        // Return any ETH left into this contract
+        // (can happen when more ETH is sent than needed for the fast withdrawals)
+        msg.sender.sendETHAndVerify(address(this).balance, gasleft());
+    }
+
+    // -- Internal --
+
+    function executeFastWithdrawal(FastWithdrawal memory fastWithdrawal)
+        internal
+    {
+        // Currently here because the compiler hangs when doing this in the constructor
+        if (DOMAIN_SEPARATOR == bytes32(0)) {
+            DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+        }
+
+        // Compute the hash
+        bytes32 hash = EIP712.hashPacked(
+            DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encodePacked(
+                    FASTWITHDRAWAL_TYPEHASH,
+                    fastWithdrawal.exchange,
+                    fastWithdrawal.from,
+                    fastWithdrawal.to,
+                    fastWithdrawal.token,
+                    fastWithdrawal.amount,
+                    fastWithdrawal.feeToken,
+                    fastWithdrawal.fee,
+                    fastWithdrawal.onchainFeePayment,
+                    fastWithdrawal.validUntil
+                )
+            )
+        );
+
+        // Make sure the tx isn't used multiple times
+        require(txHash[hash] == false, "TX_REPLAY");
+        txHash[hash] = true;
+
+        // Check the signature
+        require(hash.verifySignature(fastWithdrawal.from, fastWithdrawal.signature), "INVALID_SIGNATURE");
+
+        // Check the time limit
+        require(now <= fastWithdrawal.validUntil, "TX_EXPIRED");
+
+        // Transfer the tokens to the requested address
+        // using funds from `msg.sender`.
+        transfer(
+            fastWithdrawal.to,
+            fastWithdrawal.token,
+            fastWithdrawal.amount
+        );
+
+        // If the fee is paid offchain in the same token as the main transfer
+        // just add it to the transfer amount
+        if (!fastWithdrawal.onchainFeePayment && fastWithdrawal.token == fastWithdrawal.feeToken) {
+            fastWithdrawal.amount = fastWithdrawal.amount.add(fastWithdrawal.fee);
+            fastWithdrawal.fee = 0;
+        }
+
+        IExchangeV3 exchange = IExchangeV3(fastWithdrawal.exchange);
+
+        // Approve the offchain transfer from the withdrawing account back to the mediator (msg.sender)
+        exchange.approveConditionalTransfer(
+            fastWithdrawal.from,
+            msg.sender,
+            fastWithdrawal.token,
+            fastWithdrawal.amount
+        );
+
+        // Fee payment to the mediator (msg.sender)
+        if (fastWithdrawal.fee > 0) {
+            if (fastWithdrawal.onchainFeePayment) {
```

Comment:
It's more efficient to do the fee payment as part of the single approved transfer, so even if it is possible to do it onchain it could be less efficient which would be worse for all parties involved.

But this is really just an example implementation which I think makes sense. It will depend on how exchanges want to provide this service.